Fix min_weight_matching to convert edge weights without reciprocal

dschult opened the pull request:
Fixes #5388 

Add test and then fix code and docs.

Discussion point:  I think we need to remove the `maxcardinality` parameter because it doesn't make sense in a minimum weight matching to allow edges to not be included. You would return a matching with no edges. That would certainly minimize the edge weights.  Inside `min_weight_matching` we should always call `max_weight_matching` with `maxcardinality=True`.

I think this keyword should just be removed. But I deprecated it -- only instead of a warning it raises a NetworkXError telling the user that the concept doesn't make sense and that the parameter would be removed in v3.  But I still think it should just be removed now -- with a note in the release notes.

Thoughts?

mjschwenne commented:
Reading about this bug reminded me of #4836 since they are both a similar issue with converting a maximal algorithm to take the minimum. While I haven't had the time yet to work on that issue, it is still on my mind. I wonder if other functions which take a minimal approach using a maximal algorithm have this issue. It might be a good idea to check if these are part of a large problem. I *should* start to get some free time in the coming week and try to compile a list to test.

rossbar reviewed (commented):
LGTM thanks @dschult - just a couple docstring nits!

Also I think this should include a release note and a bullet in `doc/developer/deprecations` so we don't forget to remove the parameter for 3.0!

dschult commented:
Sorry this took so long...  I've been worried whether adding 1 to each edge weight to avoid an edge with zero weight would change the max_weight_matching by favoring matchings with more edges in them. It took me a little bit to figure out that all the matchings that we compare in `max_weight_matching` using option `maxcardinality=True` have the same number of edges (cardinality). So adding a constant to the edge weights doesn't affect which matching is optimal. I switched from adding 1e-6 to adding 1 because it would allow integer edge weights to still be integer values. 

I've fixed the documentation errors/suggestions, so I think this is ready for re-review.

rossbar commented on .pre-commit-config.yaml:
```diff
@@ -7,7 +7,7 @@ repos:
     - id: black
 -   repo: https://github.com/asottile/pyupgrade
-    rev: v2.29.1
+    rev: v2.31.1
```
I'm not sure if this should get squashed into this PR or might be better to have in its own or as part of a general maintenance PR. Not a big deal, just thinking out loud!

rossbar reviewed (approved):
LGTM, thanks @dschult !

mjschwenne reviewed (approved):
Docstring explanations about the new method was informative and everything looks good!

rossbar commented:
Thanks @dschult & @mjschwenne !

mjschwenne commented:
Any reason your approval is green and mine is gray?

dschult commented:
[Looks like](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews)  green checks are from people with write access to the repository while gray checks are from people without write access.  Is the difference in write access intentional or how is it determined?

MridulS commented:
I think it's because @mjschwenne is currently part of `triage` team, and that team can move around issues, close them but they can't merge PRs into the `main` branch.
